endpoints-framework/src/main/java/com/google/api/server/spi/ServiceException.java

```diff
@@ -102,6 +115,49 @@ public String getDomain() {
     return null;
   }
 
+  /**
+   * Adds a new extra field.
+   * @return this
+   * @throws IllegalArgumentException if {@code key} is one of the reserved keyword {@link #EXTRA_FIELDS_RESERVED_KEYWORDS}.
+   */
+  public ServiceException addExtraField(String key, String value) {
+    return addExtraFieldInternal(key, value);
+  }
+
+  /**
+  /**
+   * Adds a new extra field.
+   * @return this
+   * @throws IllegalArgumentException if {@code key} is one of the reserved keyword {@link #EXTRA_FIELDS_RESERVED_KEYWORDS}.
+   */
+  public ServiceException addExtraField(String key, Boolean value) {
+    return addExtraFieldInternal(key, value);
+  }
+
+  /**
+   * Adds a new extra field.
+   * @return this
+   * @throws IllegalArgumentException if {@code key} is one of the reserved keyword {@link #EXTRA_FIELDS_RESERVED_KEYWORDS}.
+   */
+  public ServiceException addExtraField(String key, Number value) {
+    return addExtraFieldInternal(key, value);
+  }
+
+  private ServiceException addExtraFieldInternal(String key, Object value) {
+    Preconditions.checkNotNull(key);
+    Preconditions.checkArgument(!EXTRA_FIELDS_RESERVED_KEYWORDS.contains(key), "The keyword '%s' is reserved", key);
+    Object previousValue = extraFields.put(key, value);
+    if (previousValue != null) {
+      logger.atFine().log("Replaced extra field %s: %s => %s", key, previousValue, value);
+    }
+    return this;
+  }
+
+  @Nonnull
```

Comment:
I would not put @Nonnull here, as it's not used anywhere in the framework.
But I would add Javadoc with a note about nullity.